Client/NoOpRunner.Client/Rendering/GameWindowRenderer.cs

```diff
@@ -1,48 +1,188 @@
-﻿using NoOpRunner.Core;
+﻿using System;
+using NoOpRunner.Core;
 using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
 using System.Windows.Controls;
 using System.Windows.Media;
+using System.Windows.Media.Animation;
+using System.Windows.Media.Imaging;
 using System.Windows.Shapes;
+using NoOpRunner.Core.Shapes;
 
 namespace NoOpRunner.Client.Logic.Rendering
 {
     public static class GameWindowRenderer
     {
-        public static void Render(GameWindow window, Canvas canvas)
+        // private static AnimationRenderer _animationRenderer;
+        private static UIElement lastElement;
+
+        public static void Render(GameWindow window, Canvas canvas) //Rename:RenderMap???
         {
             var rectangleWidth = canvas.ActualWidth / window.SizeX;
             var rectangleHeight = canvas.ActualHeight / window.SizeY;
 
-            canvas.Children.Clear();
 
-            foreach (var pixel in window.GetCurrentWindow())
+            var pixels = window.GetCurrentWindowEnumerable().ToList();
+            if (canvas.Children.Count != pixels.Count)
             {
-                var rec = new Rectangle
+                canvas.Children.Clear();
+
+                var imageBrush = InitImageBrush(SpritesUriHandler.GetPlatformUri());
+
+                foreach (var pixel in pixels)
                 {
-                    Width = rectangleWidth,
-                    Height = rectangleHeight,
-                    Fill = ColorBrushMap[pixel.Color]
-                };
+                    var rec = new Rectangle
+                    {
+                        Width = rectangleWidth,
+                        Height = rectangleHeight,
+                        Fill = imageBrush
+                    };
+                    Canvas.SetLeft(rec, rectangleWidth * pixel.X);
+                    Canvas.SetBottom(rec, rectangleHeight * pixel.Y);
+
+                    canvas.Children.Add(rec);
+                }
+            }
+            else
+            {
+                for (int i = 0; i < pixels.Count; i++)
+                {
+                    canvas.Children[i].SetValue(Canvas.WidthProperty, rectangleWidth);
+                    canvas.Children[i].SetValue(Canvas.HeightProperty, rectangleHeight);
+
+                    Canvas.SetLeft(canvas.Children[i], rectangleWidth * pixels[i].X);
+                    Canvas.SetBottom(canvas.Children[i], rectangleHeight * pixels[i].Y);
+                }
+            }
 
-                Canvas.SetLeft(rec, rectangleWidth * pixel.X);
-                Canvas.SetBottom(rec, rectangleHeight * pixel.Y);
+            #region Old code
 
-                rec.MouseLeftButtonDown += (s, e) =>
+// this still retarded, why generate empty pixels?? make canvas background color
+            // var pixels = window.GetCurrentWindow();
```

Comment:
Don't leave commented code like this, if you replace it, then remove the old implementations.